### ISSUE TITLE
Custom queryset Manager with a filter by ForeignKey

### ISSUE THREAD

**Issue by dhbahr** [django-salesforce/django-salesforce#253]
I know this is very weird, and I have the feeling it *might* be related to my specific setup, but basically I'm running into an issue when attempting to invoke the `save` function it fails not being able to assert the query is a `SalesforceQuery`. 

The weird think is that it only happens with contacts. Other SalesforceModels are properly saved without any hustle.

There are no hard relationships between the contact object and any non-salesforce objects. I basically use it as a "profile" for local users, matching by email address.

Is there anything known that may lead to this?

Here is what I think is the relevant code ..

```
from salesforce import models as sfmodels
from authtools import models as atmodels
from salesforce.backend.manager import SalesforceManager


class CustomManager(SalesforceManager):

    def get_queryset(self, _alias=None):
        return super().get_queryset(_alias=_alias).filter(record_type__developer_name='Custom',
                                                          portal_status__in=('Invited', 'Active'))
...
class Contact(sfmodels.Model):
    ... # regular contact fields in here
    objects = CustomManager()

class User(atmodels.AbstractEmailUser):
    @property
    def srecord(self):
        return Contact.objects.get(email=self.email)
```

and a sample snippet:
```
>>> from authentication.models import Contact
>>> c = Contact.objects.get(email="dhbahr@gmail.com")
>>> c.portal_requires_token = True
>>> c.save()
Query
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File ".env/lib/python3.8/site-packages/django/db/models/base.py", line 740, in save
    self.save_base(using=using, force_insert=force_insert,
  File ".env/lib/python3.8/site-packages/django/db/models/base.py", line 777, in save_base
    updated = self._save_table(
  File ".env/lib/python3.8/site-packages/django/db/models/base.py", line 850, in _save_table
    updated = self._do_update(base_qs, using, pk_val, values, update_fields,
  File ".env/lib/python3.8/site-packages/django/db/models/base.py", line 900, in _do_update
    return filtered._update(values) > 0
  File ".env/lib/python3.8/site-packages/django/db/models/query.py", line 760, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File ".env/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1429, in execute_sql
    cursor = super().execute_sql(result_type)
  File ".env/lib/python3.8/site-packages/salesforce/backend/compiler.py", line 89, in execute_sql
    cursor.execute(sql, params)
  File ".env/lib/python3.8/site-packages/salesforce/backend/utils.py", line 183, in execute
    response = self.execute_django(q, args)
  File ".env/lib/python3.8/site-packages/salesforce/backend/utils.py", line 232, in execute_django
    response = self.execute_update(self.query)
  File ".env/lib/python3.8/site-packages/salesforce/backend/utils.py", line 328, in execute_update
    pks = self.get_pks_from_query(query)
  File ".env/lib/python3.8/site-packages/salesforce/backend/utils.py", line 308, in get_pks_from_query
    assert isinstance(pks, Query) and type(pks).__name__ == 'SalesforceQuery'
AssertionError
```
(Hynekcer: edited the long paths before ".env/". Django is 2.2)

**Comment by hynekcer** [comment]
Excuse, the version of Django-salesforce is not known. (It looks like a modified source code. I even can not identify an original commit with these line numbers. No problem if you send a link to a clone of django-salesforce)

Add a line `print(pks)` before assert to see what is a value before AssertionError. That helps much.

Do you use more managers for Contact? (Is is a reason why you specify an explicit manager? "objects=...")

Print `type(c)` in your example. It should look like `<class 'your_app.models.Contact'>`. (not a Contact encapsulated by a decorator or something.

Do you see some weird decorator that could cause

**Comment by dhbahr** [comment]
Hi @hynekcer thanks for the prompt reply. It does seem I had a line off on some of the methods there, probably added some `print` and removed them wrongly at some point while debugging. I have now uninstalled/reinstalled `django-salesforce` (version 0.9) and still get the same errors.

I use a custom manager because I do filter the queryset (I have updated the original snippet to better reflect this).

Here is the updated snippet including the 2 requested actions:

```

**Comment by hynekcer** [comment]
Please, can you confirm that the SQL line with ... INNER JOIN "RecordType" ... is related to your custom manager?

That custom manager is probably the first manager in the model, the "[default_manager](https://docs.djangoproject.com/en/3.0/topics/db/managers/#default-managers)".

Would be it fixed if you change the order of managers? (What is an unwanted consequence for you if the order of managers is changed? It is easier for you before I start to write a test case.)

**Comment by hynekcer** [comment]
The problem can not be reproduced. I wrote a minimalistic test, that it works. You can extend it to something small that can be copied to a new virtual env and that fails, Please use only standard salesforce objects with standard fields to demonstrate it if possible.
```
print(type(pks))     # interesting info
```

**Comment by dhbahr** [comment]
@hynekcer I would assume the inner join **is** related to the manager (I have only one on Contact), you can see it on the original post as I updated it yesterday. 
The manager basically makes sure we only deal with the specific contacts we want to allow access to the portal.
I will test removing the manager and see if that goes through and let you know.

Thanks for the insights.

**Comment by dhbahr** [comment]
@hynekcer I think I have narrowed it down.

The issue is only reproducible when the Manager includes the record id in the queryset.

Thanks for your time.

**Comment by hynekcer** [comment]
I renamed the issue.

I confirm that even an individual queryset with a filter by a foreign key attribute doesn't work currently with update() or delete() methods. I'm not sure if it is a exactly bug or only something unsupported that has an easy and even a better alternative. It it better to not support more uncertain cases than to update incorrect rows. That would be really a bug. I fix it however by a clear error message at the assert command.

Docs about [Default managers](https://docs.djangoproject.com/en/3.0/topics/db/managers/#default-managers)
> ... it’s a good idea to be careful in your choice of default manager in order to avoid a situation where overriding get_queryset() results in an inability to retrieve objects you’d like to work with.

It is important especially with Salesforce, because a `.save()` method with your default manager is compiled even with a normal database by Django to

    UPDATE table_name SET column1 = value1, ...
    WHERE id=... AND PortalStatus=... AND related_table....

It verifies again that all conditions are valid and maybe if you change e.g. PortalStatus then the object will be not saved(), but it is still only one SQL

With Salesforce, if the queryset method of default manager contains a filter then SQL query is run to get filtered primary keys and then a non-SQL command is run to update the object or objects. That is not optimal. 

1. You can try to work with a renamed default manager and your manager as objects:

    `default_manager = SalesforceManager()`  # the first manager, the name is not important  
    `objects = ... your manager`

2. Cache the `record_type_id` in come variable and filter by it instead of using a join so frequently.
    `...filter(record_type_id=...)`

**Comment by hynekcer** [comment]
The rest of the issue is fixed by a clear error message  
`AssertionError: Too complicated queryset.update(). Rewrite it by two querysets. See docs wiki/error-messages`

It can be searched in the wiki [AssertionError: Too complicated queryset.update()](https://github.com/django-salesforce/django-salesforce/wiki/Error-messages/_edit#q-assertionerror-too-complicated-querysetupdate)

Because the suggestion in the wiki seems that it solves the rest your problem completely, I'm closing it.

**Comment by dhbahr** [comment]
Thanks @hynekcer !

**Comment by hynekcer** [comment]
The bug is found and will be easily fixed in the next release. A base manager did depend on the default manager. An explicit base manager is now assigned in the SalesforceModel class, independent on the default manager. The method `.save()` is executed now again by one HTTP request, as expected.

**Comment by hynekcer** [comment]
Fixed by PR #258.
